### PR TITLE
5981 non modal pitch and speed dialog

### LINKED ISSUE #5944: Cancelling the Pitch & Speed dialog by ESC doesn't restore the previous pitch value and submits the entered value instead
### Bug description

_No response_

### Steps to reproduce

1. Have a clip
2. Navigate to the Pitch and Speed dialog
3. Enter some value in the Clip Pitch and/or Cents text box
4. Press ESC on the keyboard
5. Observe

### Expected behavior

The value entered in step 3 has not been submitted

### Actual behavior

The value entered in step 3 has been submitted

### Audacity Version

current master / alpha / nightly build

### Operating system

macOS

### Additional context

_No response_

### LINKED ISSUE #5981: Non-modal pitch-and-speed dialog
### Your idea

Pitch and speed dialog isn't modal anymore, meaning having it open doesn't block interaction with the rest of the UI.

### Problem to be solved

Changing speed isn't allowed (yet) during playback. So if one wants to change speed, that person has to close the dialog, press stop, and then open it again.

### Prior art

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #5981 

- [ ] I signed [CLA](https://www.audacityteam.org/cla/)
- [ ] The title of the pull request describes an issue it addresses
- [ ] If changes are extensive, then there is a sequence of easily reviewable commits
- [ ] Each commit's message describes its purpose and effects
- [ ] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

QA:
- [x] While the P&S dialog is open, selecting another clip updates the title in the dialog. ("Selecting" means clicking either the clip's waveform or affordance bar or tabbing through the clips of a track.)
- [x] Ctrl+Shift+P opens dialog for last selected clip
- [x] History entries are added sensibly.
- [x] The lowest possible speed value is bound by the beginning of the following clip (preventing overlap)
- [x] The close button or pressing ESC close the dialog.
- [x] Deleting the clip that's represented in the dialog closes the dialog.
- [x] The dialog displays the name of the last clicked clip.
- [x] Changing the pitch value in the dialog updates the value displayed on the clip.
- [x] Starting playback/recording play greys out the speed box, which gets re-activated when playback/recording stops.
- [x] Playback start/stop events of one project doesn't activate/deactivate speed box in other project.
- [x] When clicking the arrow icon of a clip, the dialog still opens with focus on the semitone box, and when clicking the clock icon, on the speed box.
- [x] Accessibility still works
- [x] Changing speed with stretch handles updates the value in the dialog
- [x] Changing speed by changing project tempo updates the value in the dialog
- [x] Changing pitch semitone and cent values behaves as gracefully as before when going from positive to negative values. (Pardon me for not elaborating the subtleties, I think you'll notice if there is a difference in behaviour.)
- [x] Clicking on the right channel of a stereo clip behaves as when clicking on the left channel.
- [x] Maybe this resolves #5944 ?

### CODE REVIEW

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @saintmatthieu 's branch build for this.

This looks properly non-modal to me.

UI can invoke the speed&[itch dialog from the Clip-handle and then still press Play or other buttons in the app.

**Comment by saintmatthieu** [comment]
> Testing on W10 with @saintmatthieu 's branch build for this.
> 
> This looks properly non-modal to me.
> 
> I can invoke the speed&pitch dialog from the Clip-handle and then still press Play or other buttons in the app.

Thanks @petersampsonaudacity ! When you tested the focus selection (pitch or speed depending on why clip button one clicked, the arrow or the clock) was broken.

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @saintmatthieu 's latest branch build for this:  audacity-win-3.5.0-alpha-20240315+de007d8-x64

The arrow amd the clock in the Clip-handle bow mirror the setting in the change Pitch&Speed dialog
image

I also note that nocely in this non-modal dialog
1. when you are not playing the speed change is not grayed-out and can be used
2. Press Play and observe the Change Speed is grayed-out and inoperable
3. Press Pause and it remains grayed-out
4. Press Stop and observe that it becomes non-grayed out and usable again

**Comment by DavidBailes** [comment]
@saintmatthieu . A couple of issues:
A. At least on Windows, it's standard for a modeless dialog to just have a close button, rather than OK and Cancel buttons. That is, all changes take place immediately, and you can't cancel any changes. With the current behaviour, a user can:
1. Open the pitch and speed dialog for a clip
2. In the main windows, press alt + up arrow a few times to increase the pitch.
3. In the pitch and speed dialog, press Cancel. This undoes actions taken in the main window, which I don't think should happen, and which I don't think the user would expect. Cancel should only undo changes made in the Dialog.

Just having a Close button removes this issue (and maybe others).

B. With the full set of default keystrokes, there are the keystrokes Alt + F6 and Shift + Alt + F6 which cycle you round the main window and any modeless dialogs, either forwards or backwards respectively. There is a bug for the pitch and speed dialog:
With the full set of default keystrokes:
1. Open the pitch and speed dialog for a clip.
2. Press Alt + F6. The main window becomes the focus.
3. Press Alt + F6. The main window remains the focus. The pitch and speed dialog should become the focus again.

**Comment by LWinterberg** [comment]
> Just having a Close button removes this issue (and maybe others).

A close button wouldn't work as the stretching can fail still. Applying stretching a real time is not great either, as typing in a new one-digit number would cause extreme stretches which are almost guaranteed to bump into something.

**Comment by DavidBailes** [comment]
> A close button wouldn't work as the stretching can fail still. Applying stretching a real time is not great either, as typing in a new one-digit number would cause extreme stretches which are almost guaranteed to bump into something.

Then shouldn't it be  a modal dialog?. (If it's got OK and Cancel buttons, users will think it's a modal dialog.)

**Comment by petersampsonaudacity** [comment]
> Then shouldn't it be a modal dialog?.

@DavidBailes  - it looks like a modal dialog to me, it blocks interaction with the rest of Audacity until this dialog is dealt with.

**Comment by DavidBailes** [comment]
> Anyways, please try https://github.com/audacity/audacity/actions/runs/8299522237

@saintmatthieu @LWinterberg For the speed, I think a user would probably also want to be able to type in a value. In addition, in previous commits a user has been able to enter values with decimal places, eg 89.6, which they can no longer do.
There would also be an accessibility issue to sort out, as the control is read as an editable text box (which it obviously isn't), but I think there first needs to be a decision about the usability of this control for users in general.

**Comment by DavidBailes** [comment]
> @DavidBailes - it looks like a modal dialog to me, it blocks interaction with the rest of Audacity until this dialog is dealt with _(it does when I test it)_.

@petersampsonaudacity . Could you give me the commit number of the build that you're referring to (there have been a few recently)?

**Comment by saintmatthieu** [comment]
> > Anyways, please try https://github.com/audacity/audacity/actions/runs/8299522237
> 
> @saintmatthieu @LWinterberg For the speed, I think a user would probably also want to be able to type in a value. In addition, in previous commits a user has been able to enter values with decimal places, eg 89.6, which they can no longer do. There would also be an accessibility issue to sort out, as the control is read as an editable text box (which it obviously isn't), but I think there first needs to be a decision about the usability of this control for users in general.

I can do everything on my Windows machine (enter value directly also with decimal places) so long as I do not _clear_ the text before, which is obviously a bug. Looking into this.

About the reader seeing this as a text box: because we need decimal value support, a standard `wxSpinCtrl` object doesn't do it. I needed to re-use the spin-control wrapper already used for the project tempo. Was this concern of yours, @DavidBailes , somehow eliminated for project tempo?

**Comment by saintmatthieu** [comment]
Maybe what I should add is a call such as
```cpp
#if wxUSE_ACCESSIBILITY
            txtCtrl->SetAccessible(safenew WindowAccessible(txtCtrl));
#endif
```
? Then the narrator reads "Clip Speed pane, speed % colon" instead of "Control pane, speed % colon".

**Comment by petersampsonaudacity** [comment]
> @petersampsonaudacity . Could you give me the commit number of the build that you're referring to (there have been a few recently)?

@DavidBailes 
audacity-win-3.5.0-alpha-20240315+4a1df2d-x64

**Comment by DavidBailes** [comment]
> > @petersampsonaudacity . Could you give me the commit number of the build that you're referring to (there have been a few recently)?
> 
> @DavidBailes audacity-win-3.5.0-alpha-20240315+4a1df2d-x64
That gives a modeless dialog here:
image

**Comment by petersampsonaudacity** [comment]
@DavidBailes 

image

Works fine for me - I can
1. invoke the dialog from the clip handle - and with it onscreen
2. Press Play (or Pause or Stop)
3. interact with the volume slider in the meter toiolbar and the pan & gain sliders in the TCP
4. and as it is Playing I can dynamically change the pitch in the dialog (but not the speed - that gets grayed out)

**Comment by DavidBailes** [comment]
> I can do everything on my Windows machine (enter value directly also with decimal places) so long as I do not _clear_ the text before, which is obviously a bug. Looking into this.

I was running into that bug.
> 
> About the reader seeing this as a text box: because we need decimal value support, a standard `wxSpinCtrl` object doesn't do it. I needed to re-use the spin-control wrapper already used for the project tempo. Was this concern of yours, @DavidBailes , somehow eliminated for project tempo?

It's not a problem a screen reader reading the control as an editable text box, as long as the user can edit the text!

**Comment by DavidBailes** [comment]
> Maybe what I should add is a call such as
> 
> ```c++
> #if wxUSE_ACCESSIBILITY
>             txtCtrl->SetAccessible(safenew WindowAccessible(txtCtrl));
> #endif
> ```
> 
> ? Then the narrator reads "Clip Speed pane, speed % colon" instead of "Control pane, speed % colon".

The SpinControl contains a text box, and it's the latter which is the focus when the control is the focus.
It's only narrator that by default also reads the name of the containing window when it reads the name of a focused control. Occasionally that name is useful but often it isn't. In this case it isn't useful to have any name, and it would probably be better for narrator just to read something like "pane, speed % colon". If an empty name was going to be set for the containing SpinControl it would probably be better if this was done for the class, rather than having to do it for individual instances. I don't think it's an important issue at the moment. (If you really wanted to implement it for this instance, then you could set an accessible as above, and in PitchAndSpeedDialog::PopulateOrExchange() change .Name(XO("Clip Speed")) to set an empty string instead.)

Also note that the text box has a custom accessible object. This means that the accessible name can be different from the name before the control. So if you wanted you could have the word "speed" before the control, and "%" after the control.
In ShuttleGuiBase::AddSpinControl() in the constructor for the SpinControl, you set the prompt to be the accessibility name.
You could override that in PitchAndSpeedDialog::PopulateOrExchange() by calling SetName() on the textCtrl with the string "speed %"

In passing, you currently can't close the dialog by pressing Enter or Esc.

**Comment by DavidBailes** [comment]
> Works fine for me - I can
> 
> 1. invoke the dialog from the clip handle - and with it onscreen
> 2. Press Play (or Pause or Stop)
> 3. interact with the volume slider in the Meter toolbar and the pan & gain sliders in the TCP
> 4. and as it is Playing I can dynamically change the pitch in the dialog _(but not the speed - that gets grayed out as you can se in the image above)_

But you said previously "it looks like a modal dialog to me, it blocks interaction with the rest of Audacity until this dialog is dealt with (it does when I test it)".

**Comment by DavidBailes** [comment]
> I believe that was true at the time of writing and with @saintmatthieu 's branch as it was at the time.
> 
> But his latest branch is definitely non-modal when I test it

Yes.

**Comment by saintmatthieu** [comment]
@DavidBailes I briefly explored your suggestions, but couldn't get to the result where "%" is read after the value. Feel free to push a commit on this branch to perfect accessibility.

**Comment by DavidBailes** [comment]
> @DavidBailes I briefly explored your suggestions, but couldn't get to the result where "%" is read after the value. Feel free to push a commit on this branch to perfect accessibility.

@saintmatthieu . The suggestion wasn't to get the "%" read by screen readers after the value. Rather it was for the "%" to appear after the value for sighted users. But I think it's good enough now, so it's only something that I might look at in the (distant) future after your commit has been merged.

File: libraries/lib-audio-io/AudioIO.cpp [hunk 1]
```diff
@@ -1670,6 +1669,8 @@ void AudioIO::StopStream()
             false });
    }
 
+   ResetOwningProject();
```
[the remark this answers is not in the document]
Comment (saintmatthieu) [reply libraries/lib-audio-io/AudioIO.cpp:1672]: Without this change, the `AudioIOEvent::PLAYBACK` callback is called with an empty project pointer, and so the callee cannot check the playback of which project it is that stopped.
Indeed, you identified a place where this change could affect behaviour. I examined other places subscriptions of this callback and this seems to be the only one (besides PitchAndSpeedDialog) that cares about the value of `event.pProject` when `!event.on`. I'll push a fix for `MeterPanel`.

File: src/ProjectManager.cpp [hunk 2]
```diff
@@ -565,6 +566,8 @@ void ProjectManager::OnCloseWindow(wxCloseEvent & event)
    // to save the state of the toolbars.
    ToolManager::Get( project ).Destroy();
 
+   PitchAndSpeedDialog::Destroy(project);
```
[the remark this answers is not in the document]
Comment (saintmatthieu) [reply src/ProjectManager.cpp:569]: You mean `wxEVT_CLOSE_WINDOW` on `this` ? This is triggered when I close the dialog, meanwhile `Destroy` I think is meant for when closing the project. I just tried, though, and it freezes Audacity.

[the remark this answers is not in the document]
Comment (saintmatthieu) [reply src/ProjectManager.cpp]: As discussed, used `ProjectWindowDestroyedMessage` - thanks for the suggestion, it's better now.

**Review by vsverchinsky — APPROVED** [review]

**Comment by saintmatthieu** [comment]
@DavidBailes, I realize something you recommended still hasn't been addressed, namely
> B. With the full set of default keystrokes, there are the keystrokes Alt + F6 and Shift + Alt + F6 which cycle you round the main window and any modeless dialogs, either forwards or backwards respectively. There is a bug for the pitch and speed dialog:
With the full set of default keystrokes:
...
I'd like to have this for my own use, too, but haven't been able to change scope that way even with existing windows such as the history window. Is this feature really still working properly?

**Comment by DavidBailes** [comment]
> I'd like to have this for my own use, too, but haven't been able to change scope that way even with existing windows such as the history window. Is this feature really still working properly?

The Alt + F6 and Shift + Alt + F6 keystrokes are only available by default when using the full set of default shortcuts. (In the shortcuts page of preferences, press the defaults button and choose full.)
With the full set, I think you'll find that you can use these shortcuts to move back and forth to the history window (and all the real time effect dialogs).

**Pull request merged**